Replace the MVP with MVVM design pattern for Product Detail

0nko opened the pull request:
This PR replaces the MVP for MVVM design pattern for the Product Detail screen. The base uses the `ViewModel` and `LiveData` from the Architecture components. 

There is a new `ProductRepository`, which wraps the FluxC event for product-fetching and makes use of Kotlin coroutines.

I've added a new app-layer model `Product` that uses appropriate types for convenience, particularly for currency values (`BigDecimal`), dates (`Date`) and some custom enumerations (using sealed classes and enums).

There are a bunch of utility and helper classes for ViewModels, LiveData and unit testing, which simplifies some work adds some coding standards from WPAndroid.

The unit tests have been updated to test the VM. This required an update of the Mockito Kotlin library, because only the newer version supports the testing of coroutines. An AssertJ and androidx.arch.core:core-testing library were also added.

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/di/ViewModelModule.kt:
```diff
@@ -0,0 +1,22 @@
+    @Binds
+    @IntoMap
+    @ViewModelKey(ProductDetailViewModel::class)
+    internal abstract fun pluginOrderDetailViewModel(viewModel: ProductDetailViewModel): ViewModel
```
np: since this is product detail view model could we change the name of this method?

0nko replied:
You're right! This was a leftover from the initial Order detail screen work.

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailFragment.kt:
```diff
@@ -39,31 +42,36 @@ import com.woocommerce.android.util.StringUtils
+    @Inject lateinit var viewModelFactory: ViewModelProvider.Factory
+
     @Inject lateinit var uiMessageResolver: UIMessageResolver
     @Inject lateinit var networkStatus: NetworkStatus
```
np: Not sure if we need the `NetworkStatus` injected here anymore

0nko replied:
You're right, good catch!

anitaa1990 commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductStockStatus.kt:
```diff
@@ -0,0 +1,24 @@
+import androidx.annotation.StringRes
+import com.woocommerce.android.R
+
+sealed class ProductStockStatus(@StringRes val stringResource: Int = 0, val value: String = "") {
```
Love that the logic to get the string id from the product `status` has been moved into its own class <3

0nko commented:
@anitaa1990, thank you for the review! I addressed all of the issues above.

I totally forgot about the UI tests and I added the missing mock modules. While I was working on these changes I also refactored the `ProductDetailViewModel` and removed all public methods for getting the currency and dimensions. This is not the correct way for the view to access data, since everything should be passed on through the `LiveData`. Now any logic pertaining to the formatting is contained within the ViewModel and the fragment just gets what it needs, ready for display.

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/utility/ResourceProvider.kt:
```diff
@@ -0,0 +1,29 @@
+import javax.inject.Singleton
+
+@Singleton
+class ResourceProvider @Inject constructor(private val context: Context) {
```
Yeah, I copied them over from WPAndroid when I started because they are pretty common in VMs, but let's remove them until they're needed.

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/utility/LiveDataUtils.kt:
```diff
@@ -0,0 +1,277 @@
+import androidx.lifecycle.Observer
+import androidx.lifecycle.Transformations
+import kotlinx.coroutines.CoroutineScope
+
```
Ok, I got rid of them.

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailFragment.kt:
```diff
@@ -222,26 +257,26 @@ class ProductDetailFragment : BaseFragment(), ProductDetailContract.View, Reques
-        val hasPricingInfo = product.price.isNotEmpty() ||
-                product.salePrice.isNotEmpty() ||
-                product.taxClass.isNotEmpty()
+        val hasPricingInfo = product.price != null && product.salePrice != null || product.taxClass.isNotEmpty()
```
Agh, good catch! I probably replaced the `||` with `&&` by accident..

0nko commented on WooCommerce/src/main/kotlin/com/woocommerce/android/viewmodel/utility/ThrottleLiveData.kt:
```diff
@@ -0,0 +1,35 @@
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+class ThrottleLiveData<T> constructor(
```
I removed the class, since it was being used.

0nko commented:
@AmandaRiu, thank you for the thorough review!

> One thing I noticed is that the ProductDetailViewModel looks to be written so it can be reused for different products, but a new version is created every time the product detail view is opened. Is this as intended?

I think you're assuming that by seeing the `start(productId)`. That's actually to just initialize it. Because the ViewModel is injected, we can't supply the parameters known at run-time. The actual lifecycle of a VM is tied to the underlying fragment or activity (which is what we want, so that the LiveData observers are automatically unsubscribed). So whenever a ProductDetailFragment gets started, a ProductDetailViewModel is created, as well.


> I also found the weight and dimensions of the product is not being properly formatted

I wasn't able to reproduce this fully. The decimals used to get rounded. Now if there's a float value it's not being rounded anymore. I can see that the weight and dimension units are missing on your screenshot but they're loaded fine in my stores. Do you know why that could be?

![image](https://user-images.githubusercontent.com/1522856/62254719-8e2cb000-b3fa-11e9-8e2b-541391476521.png)

AmandaRiu commented:
> I think you're assuming that by seeing the start(productId). That's actually to just initialize it. Because the ViewModel is injected, we can't supply the parameters known at run-time. The actual lifecycle of a VM is tied to the underlying fragment or activity (which is what we want, so that the LiveData observers are automatically unsubscribed). So whenever a ProductDetailFragment gets started, a ProductDetailViewModel is created, as well.

Ah, that makes sense. Thank you for the thorough explanation!

> I wasn't able to reproduce this fully. The decimals used to get rounded. Now if there's a float value it's not being rounded anymore. I can see that the weight and dimension units are missing on your screenshot but they're loaded fine in my stores. Do you know why that could be?

This was just really bizarre, and I haven't really dug into why yet because it's late, but it seems the issue was that the user I was logging in with had a role of Shop Manager. As soon as I changed it to Administrator, the units started working again. It wasn't even your code, the `WCProductSettingsModel` table was missing the units configured for my site. So that's why it wasn't showing up. 

Everything else looks great! Nice work! I'm ready to merge, but since products was @nbradbury's baby, I'd really like him to do a review on this PR as well just to make sure we're not missing anything.

nbradbury commented:
@AmandaRiu I'm good to merge this, but it looks like it needs you're approval still.
